Lib/asyncio/futures.py

```diff
@@ -415,10 +425,56 @@ def wrap_future(future, *, loop=None):
     return new_future
 
 
+def future_add_to_awaited_by(fut, waiter, /):
+    """Record that `fut` is awaited on by `waiter`."""
+    # For the sake of keeping the implementation minimal and assuming
+    # that most of asyncio users use the built-in Futures and Tasks
+    # (or their subclasses), we only support native Future objects
+    # and their subclasses.
+    #
+    # Longer version: tracking requires storing the caller-callee
+    # dependency somewhere. One obvious choice is to store that
+    # information right in the future itself in a dedicated attribute.
+    # This means that we'd have to require all duck-type compatible
+    # futures to implement a specific attribute used by asyncio for
+    # the book keeping. Another solution would be to store that in
+    # a global dictionary. The downside here is that that would create
+    # strong references and any scenario where the "add" call isn't
+    # followed by a "discard" call would lead to a memory leak.
+    # Using WeakDict would resolve that issue, but would complicate
+    # the C code (_asynciomodule.c). The bottom line here is that
+    # it's not clear that all this work would be worth the effort.
+    #
+    # Note that there's an accelerated version of this function
+    # shadowing this implementation later in this file.
```

Comment:
If you call this one `future_add_to_awaited_by_py` and the accelerator one `future_add_to_awaited_by_c` and then assign one or the other to `future_add_to_awaited_by` in the try-except at the bottom, then you won't need this comment and it would avoid any potential confusion. Also probably easier to test.

Reply:
We don't do that for any other accelerated/pure-python function. They are called the same and they're disambiguated later by calling them either _CThing or _PyThing. This diff is already big as is so I'd leave that potential refactor for a different time.